### PR TITLE
[WIP] Introduce autotools

### PR DESCRIPTION
**Pull request by nhlfr**
cri-o requires several C libraries and provides options to use some of them or not. Autotools may help in managing C-based dependencies, generating CFLAGS, LIBS, LDFLAGS etc.

Using autotools instead of environment variables may be also more intuitive for packaging cri-o in future.

Signed-off-by: Michal Rostecki <michal@kinvolk.io>

---

TODO

- [x] use `--with` arguments in configure script instead of providing environment variables
- [x] use CFLAGS for the external libraries provided by autotools instead of using pkg-config manually
- [ ] look for
  - [x] glib
  - [x] seccomp
  - [x] apparmor
  - [x] libdevmapper
  - [x] btrfs-prog libs
  - [ ] static glibc
- [ ] try to get rid of warning about GNU Make extensions
- [ ] try to use autotools for running tests
- [ ] try to replace scripts from hack/ by autoconf or pkg-config checks
- [ ] adjust Travis CI jobs to work with autotools

### CODE REVIEW

**Comment by cyphar** [comment]
My concern with this is that autotools doesn't really play well with Go packages. Not to mention that `conmon` is likely going to be rewritten in Rust (which also doesn't play well with autotools) -- #297. So I'd recommend waiting to see what the other maintainers think before you waste your weekend working on something we might not merge.

**Comment by mrunalp** [comment]
I think it makes sense to use it for conmon. I have had good experience using it for oci-systemd-hook. autotools make it easier to build and package. @cyphar We could rework this to use cargo if/when we use rust. Meanwhile if we restrict this to C portions of the code, I don't see many downsides. WDYT?

**Comment by nhlfr** [comment]
I'm not very familiar with Rust (nor any build systems designed for it), but autotools aren't that bad for Go (rkt is using it), and IMO they're better than just a raw Makefile we have now.

I see that rewriting conmon to Rust will wait some time until libraries you're considering get more mature. However, if you already know any good build system which plays well with both Go and Rust, I'm open for suggestions. Maybe bazel? It looks like a good choice, but I need to take a deeper look.

**Comment by nhlfr** [comment]
Any more thoughts?

**Comment by mrunalp** [comment]
@nhlfr I am fine with us using autotools for conmon. Not sure if it makes sense for the Go portion given that it is quite simple today as we just vendor in all the dependencies in Go.
cc: @rhatdan

**Comment by jawnsy** [comment]
I don't particularly like autotools, but they're the de facto standard for C programs and libraries.  I've used it in the past when writing a C library, and woe will befall those that try to debug the thing (unless you love m4), but when it works, it works well.

Vendors like Fedora, Ubuntu, Debian, etc are familiar with it, and usually it's integrated well with build tools for those platforms (e.g. debhelper).  There are alternatives like CMake, and while I see it used by projects from time to time, autotools/libtool is by far the most common.

@cyphar makes a good point, that autotools & Go aren't well-integrated, but I guess for the Go portions we can continue using the Makefile with `go install`

For a good laugh, though, try googling "[autohell](https://www.google.com/search?q=gnu+autohell)", which includes [this rant by an old friend of mine](http://www.shlomifish.org/open-source/anti/autohell/) from the Perl community.  But these rants shouldn't prevent us from adopting it, since it's been said: "_There are only two kinds of languages: the ones people complain about and the ones nobody uses._"

File: configure.ac [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+PKG_CHECK_MODULES([glib], [glib-2.0])
+PKG_CHECK_MODULES([devmapper], [devmapper])
+AC_CHECK_HEADERS([btrfs/ioctl.h], [], [])
+
+AC_ARG_WITH([selinux],
+	    [AS_HELP_STRING([--with-selinux],
+			    [enable SELinux support])],
+			    [],
+			    [with_selinux=yes])
```
Comment (rhatdan) [inline configure.ac]:
```
+ AS_CASE(["$with_selinux"],
+	[yes], [PKG_CHECK_MODULES([libselinux], [libselinux])],
+	[no], []
+	[PKG_CHECK_MODULES([libselinux], [libselinux], [have_selinux=yes], [have_selinux=no])])
```
Should work.

**Comment by cyphar** [comment]
@jawnsy 

> I don't particularly like autotools, but they're the de facto standard for C programs and libraries.

I do agree, and I actually kinda like autotools. But please remember that currently we have plans to:

1. Rewrite `conmon` in Rust (or another language that is not C); and
2. Remove the `pause` binaary as soon as we can just pull an OCI pause image (the pause binary was a stopgap).

**Comment by runcom** [comment]
> Remove the pause binaary as soon as we can just pull an OCI pause image without needing Docker (the pause binary was a stopgap). Which actually should already be possible -- so maybe we should do that?

Let's do it

**Pull request closed without merging**